src/libopensc/card-piv.c

```diff
@@ -287,242 +546,1806 @@ struct piv_object {
+		sc_log(card->ctx,"SM BER-TLV not valid");
+		r = SC_ERROR_SM_AUTHENTICATION_FAILED;
+		goto err;
+	}
+
+	if ((asn1_sm_response[1].flags | SC_ASN1_PRESENT) == 0
+			|| (asn1_sm_response[2].flags | SC_ASN1_PRESENT) == 0) {
+		sc_log(card->ctx,"SM missing status or R-MAC");
+		 r = SC_ERROR_SM_AUTHENTICATION_FAILED;
+		 goto err;
+	}
+
+	if (status.len != 2
+			|| status.value == NULL
+			|| rmac8.len != 8
+			|| rmac8.value == NULL) {
+		sc_log(card->ctx,"SM status or R-MAC length invalid");
+		r = SC_ERROR_SM_AUTHENTICATION_FAILED;
+		goto err;
+	}
+
+	cmac_ctx = CMAC_CTX_new();
+	if (cmac_ctx == NULL) {
+		r = SC_ERROR_INTERNAL;
+		goto err;
+	}
+
+	/*  MCV is first, then BER TLV Encoded Encrypted PIV Data and Status */
+	macdatalen = status.value + status.len - sm_apdu->resp;
+
+	if (CMAC_Init(cmac_ctx, priv->sm_session.SKrmac, priv->sm_session.aes_size, (*cs->cipher_cbc)(), NULL) != 1
+			|| CMAC_Update(cmac_ctx, priv->sm_session.R_MCV, MCVlen) != 1
+			|| CMAC_Update(cmac_ctx, sm_apdu->resp, macdatalen) != 1
+			|| CMAC_Final(cmac_ctx, priv->sm_session.R_MCV, &R_MCVlen) != 1) {
+		r = SC_ERROR_SM_AUTHENTICATION_FAILED;
+		goto err;
+	}
+
+	if (memcmp(priv->sm_session.R_MCV, rmac8.value, 8) != 0) {
+		sc_log(card->ctx, "SM 8 bytes of R-MAC do not match received R-MAC");
+		r = SC_ERROR_SM_AUTHENTICATION_FAILED;
+		goto err;
+	}
+
+	
+	ed_ctx = EVP_CIPHER_CTX_new();
+	if (ed_ctx == NULL) {
+		r = SC_ERROR_INTERNAL;
+		goto err;
+	}
+
+	/* generate same IV used to encrypt response on card */
+	if (EVP_EncryptInit_ex(ed_ctx, (*cs->cipher_ecb)(), NULL, priv->sm_session.SKenc, zeros) != 1
+			|| EVP_CIPHER_CTX_set_padding(ed_ctx,0) != 1
+			|| EVP_EncryptUpdate(ed_ctx, IV, &outli, priv->sm_session.dec_counter, 16) != 1
+			|| EVP_EncryptFinal_ex(ed_ctx, discard, &outdl) != 1
+			|| outdl != 0) {
+		sc_log(card->ctx,"SM encode failed in OpenSSL");
+		piv_log_openssl(card->ctx);
+		r = SC_ERROR_SM_AUTHENTICATION_FAILED;
+		goto err;
+	}
+
+	/* some commands to not have response data */
+	if (ee.value != NULL) {
+		p = ee.value;
+		inlen = ee.len;
+		if (inlen < 17 || *p != 0x01) { /*padding indicator is required */
+			sc_log(card->ctx, "SM padding indicatior not 0x01");
+			r = SC_ERROR_SM_AUTHENTICATION_FAILED;
+			goto err;
+		}
+
+		p++; /* skip padding indicator */
+		inlen --;
+
+		if (inlen != inlen/16 * 16) {
+			sc_log(card->ctx,"SM encrypted data not multiple of 16");
+			r = SC_ERROR_SM_AUTHENTICATION_FAILED;
+			goto err;
+		}
+
+		/*
+		 * Encrypted data has 1 to 16 pad bytes, so may be 1 to 16 bytes longer 
+		 * then expected. i.e. plain->resp and resplen.So will do last block
+		 * and recombine.
+		 */
+
+		inlen -= 16;
+		if (plain->resplen < (unsigned) inlen) {
+			sc_log(card->ctx, "SM response will not fit in resp,resplen");
+			r = SC_ERROR_SM_AUTHENTICATION_FAILED;
+			goto err;
+		}
+
+		q = plain->resp;
+
+		/* first round encryptes Dec counter with zero IV, and does not save the output */
```

Comment:
Fixed in 0a3ec34276747e957ef7fb60e5d5f9e0ac787da8
 Rename dec_counter to resp_enc_counter  because dec_counter name is misleading.  
800-74-4 says: "(i.e., the IV used to encrypt the first response after successful completion of the key establishment protocol shall be generated by encrypting '80 00 00 00 00 00 00 00 00 00 00 00 00 00 00 01' with SKENC)."
and we want to use the same (encrypted) IV the card used to encrypt the response  so we to decrypt the response using the same IV.

Reply:
I do not see this comment fixed in the 0a3ec34

Reply:
The comment was changed to:
'/* first round encryptes Enc counter with zero IV, and does not save the output */'
and the following `EVP_EncryptInit_ex` is doing an encrypt with a zero IV.

Reply:
Fixed in b46b01d18